src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+    let clientAgent = 'directlinejs'
+
+    if (customAgent) {
+      clientAgent += `; ${customAgent}`
+    }
+
+    return `${DIRECT_LINE_VERSION} (${clientAgent})`;
+  }
+
+  private async refreshToken(firstCall = true, retryCount = 0) {
+    if (firstCall) {
+      setTimeout(async () => await this.refreshToken(false, 0), refreshTokenInterval);
+      return;
+    }
+
+    if (this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    await this.waitUntilOnline();
+
+    Observable.ajax({
+      method: "POST",
+      url: `${this.domain}/tokens/refresh`,
+      timeout,
+      headers: {
+        ...this.commonHeaders()
+      }
+    })
+      .subscribe(
+        (ajaxResponse) => {
+          this.token = ajaxResponse.response.token as string;
+          setTimeout(() => this.refreshToken(false, MAX_RETRY_COUNT), refreshTokenInterval);
+        },
+        (e) => {
+          if (e.status === 403 || e.status === 404) {
+            console.warn("Error refreshing token " + e.status);
+          }else{
+            if (retryCount > 0) {
+              console.warn("Error refreshing token " + e.status + " " + retryCount + " retries left");
+              this.refreshToken(false, retryCount-1);
+            }else{
+              this.token = null;
+              console.warn("Error refreshing token " + e.status + " Retries exhausted");
+            }
+          }
+        })
+  }
+
+  postActivity(activity: Activity) {
+    if (activity.type === "message" && activity.attachments && activity.attachments.length > 0)
+      return this.postMessageWithAttachments(activity);
+
+    let resp$ = Observable.create(subscriber => {
+      let request = BFSE.StreamingRequest.create('POST', '/v3/directline/conversations/' + this.conversationId + '/activities');
+      request.setBody(JSON.stringify(activity));
+      this.streamConnection.send(request)
+        .then((resp) => {
+          if (resp.statusCode !== 200) throw new Error("PostActivity returned " + resp.statusCode);
+          let numberOfStreams = resp.streams.length;
+          if (numberOfStreams !== 1) throw new Error("Expected one stream but got " + numberOfStreams)
+          resp.streams[0].readAsString().then((idString) => {
+            let idObj = JSON.parse(idString);
+            return subscriber.next(idObj.Id)
+          });
+        })
+        .catch((e) => {
+          // If there is a network issue then its handled by
+          // the disconnectionHandler. Everything else can
+          // be retried
+          console.warn(e);
+          return subscriber.error(e);
+        });
+    });
+    return resp$;
+  }
+
+  private postMessageWithAttachments(message: Message) {
+    const { attachments, ...messageWithoutAttachments } = message;
+
+    let httpContentList = [];
+    return Observable.create(subscriber => {
+      return Observable.from(attachments || [])
+        .flatMap((media: Media) =>
+          Observable.ajax({
```

Comment:
Use `fetch` and expect exceptions, e.g. `contentUrl` is not `data:...base64` but a `http://` which is not accessible, etc.

Reply:
#262.